Location-sensitive polonius prototype: endgame

lqd opened the pull request:
This PR sets up the naive location-sensitive analysis end-to-end, and replaces the location-insensitive analysis. It's roughly all the in-progress work I wanted to land for the prototype, modulo cleanups I still want to do after the holidays, or the polonius debugger, and so on.

Here, we traverse the localized constraint graph, have to deal with kills and time-traveling (👌), and record that as loan liveness for the existing scope and active loans computations.

Then the near future looks like this, especially if the 2025h1 project goal is accepted:
- gradually bringing it up to completion
- analyzing and fixing the few remaining test failures
- going over the *numerous* fixmes in this prototype (one of which is similar to a hang on one test's millions and millions of constraints)
- trying to see how to lower the impact of the lack of NLL liveness optimization on diagnostics, and their categorization of local variables and temporaries (the vast majority of blessed expectations differences), as well as the couple ICEs trying to find an NLL constraint to blame for errors.
- dealing with the theoretical weakness around kills, conflating reachability for the two TCS, etc that is described ad nauseam in the code.
- switching the compare mode to the in-tree implementation, and blessing the diagnostics 
- apart from the hang, it's not catastrophically slower on our test suite, so then we can try to enable it on CI
- checking crater, maybe trying to make it faster :3, etc.

I've tried to gradually introduce this PR's work over 4 commits, because it's kind of subtle/annoying, and Niko/I are not completely convinced yet. That one comment explaining the situation is maybe 30% of the PR 😓. Who knew that spacetime reachability and time-traveling could be mind bending. 

I kinda found this late and the impact on this part of the computation was a bit unexpected to us. A bit more care/thought will be needed here. I've described my plan in the comments though. In any case, I believe we have the current implementation is a conservative approximation that shouldn't result in unsoundness but false positives at worst. So it feels fine for now.

r? @jackh726 

---

Fixes #127628 -- which was a assertion triggered for a difference in loan computation between NLLs and the location-insensitive analysis. That doesn't exist anymore so I've removed this crash test.

lqd committed:
disable NLL liveness optimization when using polonius

in NLLs some locals are marked live at all points if one of their
regions escapes the function but that doesn't work in a flow-sensitive
setting like polonius

lqd committed:
introduce reachability to the constraint graph

lqd committed:
replace location-insensitive analysis with location-sensitive analysis

we're in in the endgame now

set up the location-sensitive analysis end to end:
- stop recording inflowing loans and loan liveness in liveness
- replace location-insensitive liveness data with live loans computed by
  reachability
- remove equivalence between polonius scopes and NLL scopes, and only
  run one scope computation

lqd committed:
handle kills in reachability

lqd committed:
deal with naive reachability weakness

it's a bit mind-bending

lqd committed:
remove location-insensitive ICE test

lqd commented:
<sub>borrow checking is easy with time travel and spacetime reachability</sub>

![image](https://github.com/user-attachments/assets/dd6e8347-5180-436e-88ff-7d0874788df1)

jackh726 reviewed (commented):
Just starting to review

lqd commented on compiler/rustc_borrowck/src/polonius/mod.rs:
```diff
@@ -113,8 +124,19 @@ impl PoloniusContext {
+            regioncx.liveness_constraints(),
+            borrow_set,
+            &localized_outlives_constraints,
+            &mut live_loans,
```
Sure. The goal was to give a mut ref from the context. It's annoying to do now because some of the regioncx data is accessible through a shared ref and we need to mutate the context or live loans, so I've left doing that to a future PR (or if we come up with a better design now that we have all the flow mapped out).

Until then, I've force pushed returning the live loans as you suggested.

lqd commented:
Rebased and marked ready to review (but jack had already started) now that #134920 has landed.

lqd commented:
(This last force push was to fix conflicts and can be ignored)

jackh726 commented on compiler/rustc_borrowck/src/dataflow.rs:
```diff
@@ -461,34 +460,26 @@ impl<'a, 'tcx> Borrows<'a, 'tcx> {
+                // The in-tree polonius analysis computes loans going out of scope using the
+                // set-of-loans model.
+                let mut polonius_prec = PoloniusOutOfScopePrecomputer::new(body, regioncx);
+                for (loan_idx, loan_data) in borrow_set.iter_enumerated() {
```
I know this is preexisting, but personally I would prefer this to be inside `PoloniusOutOScopePrecomputer::new` (you could rename to `compute` and make it return `loans_out_of_scope_at_location`.

This could be done in a separate PR (before or after this PR lands).

lqd replied:
I can do that in the next PR, sure. The structure used to be identical to the NLL scope precomputer in this function, but that was since moved out to `calculate_borrows_out_of_scope_at_location`. I’ll try to keep the symmetry with NLLs and do what you suggest there as well if possible.

lqd replied:
I pushed a commit to do this

jackh726 commented on compiler/rustc_borrowck/src/polonius/mod.rs:
```diff
@@ -98,7 +98,7 @@ impl PoloniusContext {
         &self,
         tcx: TyCtxt<'tcx>,
-        regioncx: &RegionInferenceContext<'tcx>,
+        regioncx: &mut RegionInferenceContext<'tcx>,
```
I'm looking at this. It might be a slightly bigger refactor, but I wonder if it makes sense to move `live_loans` out of RegionInferenceContext. You basically need to manually pass it anywhere you use `is_loan_live_at`, but there must *already* be some special handling wherever that is, or it would panic.

Thoughts?

lqd replied:
Sure, I can look into it for a follow up PR. What’s the goal you’re trying to achieve? 

My recollection is that only the scope precomputer accesses these, so while there’s no special handling per se, it sounds possible to pass these optional live loans when creating the `Borrows` data flow computation instead.

lqd replied:
I see. A possibility could be if we were to duplicate more of the borrowck into two instead of one handling multiple codepaths, we wouldn’t have options.

jackh726 reviewed (commented):
Some more review comments not all the way done.

lqd commented on compiler/rustc_borrowck/src/polonius/loan_liveness.rs:
```diff
@@ -0,0 +1,87 @@
+}
+
+/// Returns the outgoing edges of a given node, not its transitive closure.
+fn outgoing_edges(
```
We do add the logical edges there in https://github.com/rust-lang/rust/pull/135290/commits/da1a5ae87b24d33d618ee5671822d6c8b56e7e3b#diff-954e3381998b1d9ebf23c2136c0289705886245ef103e76127bd4a63e559e861R178

lqd replied:
I can slightly simplify it here though.

lqd replied:
It was a trivial simplification so I didn't put it in its own commit and absorbed it into https://github.com/rust-lang/rust/pull/134980/commits/877fee9543a3376c5f7171858ff42f504b828e99

lqd committed:
move out of scope precomputer code

this addresses review comments while:
- keeping the symmetry between the NLL and Polonius out of scope
  precomputers
- keeping the unstable `calculate_borrows_out_of_scope_at_location`
  function to avoid churn for consumers

bors commented:
:umbrella: The latest upstream changes (presumably #135370) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

lqd commented:
(Force pushed to fix conflicts, not to push new code that needs reviewing)

jackh726 commented on compiler/rustc_borrowck/src/polonius/loan_liveness.rs:
```diff
@@ -42,7 +54,16 @@ pub(super) fn compute_loan_liveness<'tcx>(
+                kills.get(&current_location).is_some_and(|kills| kills.contains(&loan_idx));
+
             for succ in outgoing_edges(&graph, node) {
+                // If the loan is killed at this point, it is killed _on exit_.
```
Ah, this is modified later.

jackh726 commented:
@bors r+

bors commented:
:pushpin: Commit 8ac045dd4c1aad7eca07522a4dd9c22181674728 has been approved by `jackh726`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jackh726 8ac045dd4c1aad7eca07522a4dd9c22181674728 -->
<!-- homu: {"type":"Approved","sha":"8ac045dd4c1aad7eca07522a4dd9c22181674728","approver":"jackh726","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
